src/compiled/version/Comparable.hpp

```diff
@@ -0,0 +1,197 @@
+#ifndef REZ_COMPARABLE_HPP
+#define REZ_COMPARABLE_HPP
+
+#include "Types.hpp"
+
+constexpr bool NORMAL = false;
+constexpr bool REVERSED = true;
+constexpr bool DEFAULT = NORMAL;
+
+//
+// Comparable reversible arbitrary value
+//
+// Explicit constructor allow conversion from Comparable<T, true> to Comparable<T, false>,
+// Conversion from normal and reverse comparable is not free, it requires copy.
+//
+template<typename _Typ, bool _Rev = DEFAULT> struct Comparable
+{
+    using value_type = _Typ;
+
+    // constructor
+    template<typename... Args> explicit Comparable(Args&&... args)
+        : _value{_Typ{std::forward<Args>(args)...}}
+    {
+    }
+
+    // converting constructor
+    template<bool _> explicit Comparable(const Comparable<value_type, _>& other)
+        : _value{other.Get()}
+    {
+    }
+
+    // data access
+    const value_type& Get() const REZ_NOEXCEPT { return _value; }
+    value_type& Get()  REZ_NOEXCEPT { return _value; }
+
+    // operator->
+    const value_type* operator->() const REZ_NOEXCEPT { return &Get; }
+    value_type* operator->() REZ_NOEXCEPT { return &Get(); }
+
+    // operator*
+    const value_type& operator*() const REZ_NOEXCEPT { return Get(); }
+    value_type& operator*() REZ_NOEXCEPT { return Get(); }
+
+    // operator<
+    template<bool _> bool operator<(const Comparable<value_type, _>& other ) const REZ_NOEXCEPT
+    {
+        return (Get() < other.Get());
+    }
+
+    // operator<
+    bool operator<(const value_type& other) const REZ_NOEXCEPT
+    {
+        return (Get() < other);
+    }
+
+    // operator==
+    template<bool _> bool operator==(const Comparable<value_type, _>& other ) const REZ_NOEXCEPT
+    {
+        return (Get() == other.Get());
+    }
+
+    // operator==
+    bool operator==(const value_type& other) const REZ_NOEXCEPT
+    {
+        return (Get() == other);
+    }
+
+private:
+    value_type _value;
+};
```

Comment:
`Comparable` type can be removed. In `rez` it's used as a polymorphic base, but it turns out we can replace it with lambda or functor.